### PR TITLE
Upsert dns record

### PR DESCRIPTION
**Pull request by JorritSalverda**
Allow a dns record to be updated or upserted (update existing or create otherwise)

### CODE REVIEW

File: cloudflare/cloudflare.go [hunk 1]
```diff
@@ -27,25 +26,24 @@ func New(authentication APIAuthentication) *Cloudflare {
 }
 
 func (cf *Cloudflare) getZonesByName(zoneName string) (zonesResult, error) {
```
Comment (GVdP-Travix) [inline cloudflare/cloudflare.go]: You know that you can do:
```
func (cf *Cloudflare) getZonesByName(zoneName string) (r zonesResult, err error) {
```
And the variables `r` and `err` will be pre-initialized.

File: cloudflare/cloudflare.go [hunk 2]
```diff
@@ -159,6 +146,28 @@ func (cf *Cloudflare) GetDNSRecordByDNSName(dnsName string) (DNSRecord, error) {
...
+
+	body, err := cf.restClient.Post(createDNSRecordURI, newDNSRecord, cf.authentication)
+	if err != nil {
+		return r, err
+	}
+
+	json.NewDecoder(bytes.NewReader(body)).Decode(&r)
+
+	if !r.Success {
+		return r, errors.New("cloudflare: creating dns record failed")
```
Comment (GVdP-Travix) [inline cloudflare/cloudflare.go]: Don't constantly create new errors, but define them outside of the functions.

File: cloudflare/cloudflare.go (hunk 1 above)
Comment (JorritSalverda) [inline cloudflare/cloudflare.go]: I did that first, but then changed it after reading https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters. One of the things I'd like to discuss on Friday.

File: cloudflare/cloudflare.go (hunk 2 above)
Comment (JorritSalverda) [inline cloudflare/cloudflare.go]: Let's look at that on Friday. I understand the more targeted errors you can then check for, but I'm afraid my code will be a bit littered with top-level variables.

Would it look like https://elithrar.github.io/article/http-handler-error-handling-revisited/ or https://blog.golang.org/error-handling-and-go ?

**Pull request merged**